pkg/components/linkerd/component.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func newComponent() *component {
+	return &component{
+		ControllerReplicas: 1,
+	}
+}
+
+func (c *component) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContext) hcl.Diagnostics {
+	diagnostics := hcl.Diagnostics{}
+
+	if configBody == nil {
+		return hcl.Diagnostics{}
+	}
+
+	d := gohcl.DecodeBody(*configBody, evalContext, c)
+	if d.HasErrors() {
+		return append(diagnostics, d...)
+	}
+
+	return diagnostics
+}
+
+func (c *component) RenderManifests() (map[string]string, error) {
+	// linkerd2 is the name of the upstream chart.
+	helmChart, err := util.LoadChartFromAssets("/components/linkerd2")
+	if err != nil {
+		return nil, fmt.Errorf("load chart from assets: %w", err)
+	}
+
+	if err = c.generateLinkerDCerts(); err != nil {
+		return nil, err
+	}
+
+	values, err := internaltemplate.Render(chartValuesTmpl, c)
+	if err != nil {
+		return nil, fmt.Errorf("rendering values template failed: %w", err)
+	}
+
+	values, err = mergeValuesFiles(values, valuesHA)
+	if err != nil {
+		return nil, fmt.Errorf("merging values failed: %w", err)
+	}
+
+	// Generate YAML for the Linkerd deployment.
+	renderedFiles, err := util.RenderChart(helmChart, name, c.Metadata().Namespace.Name, values)
+	if err != nil {
+		return nil, fmt.Errorf("rendering chart failed: %w", err)
+	}
+
+	return renderedFiles, nil
+}
+
+func (c *component) Metadata() components.Metadata {
+	return components.Metadata{
+		Name: name,
+		Namespace: k8sutil.Namespace{
+			Name: "linkerd",
+			Annotations: map[string]string{
+				"linkerd.io/inject": "disabled",
+			},
+			Labels: map[string]string{
+				"linkerd.io/is-control-plane":          "true",
+				"config.linkerd.io/admission-webhooks": "disabled",
+				"linkerd.io/control-plane-ns":          "linkerd",
+			},
+		},
+		Helm: components.HelmMetadata{
+			Wait: true,
+		},
+	}
+}
+
+func indent(data string, indent int) string {
+	lines := strings.Split(data, "\n")
+
+	var gap string
+
+	// Calculate the gap/indent.
+	for i := 0; i < indent; i++ {
+		gap += " "
+	}
+
+	// For each line add the gap/indent.
+	for ind := range lines {
+		lines[ind] = gap + lines[ind]
+	}
+
+	// If the last line is empty then remove the indent from it.
+	if lines[len(lines)-1] == gap {
+		lines[len(lines)-1] = ""
+	}
+
+	return strings.Join(lines, "\n")
+}
+
+func mergeValuesFiles(dst, src string) (string, error) {
+	d, err := chartutil.ReadValues([]byte(dst))
+	if err != nil {
+		return "", fmt.Errorf("could not read values from destination: %w", err)
+	}
+
+	s, err := chartutil.ReadValues([]byte(src))
+	if err != nil {
+		return "", fmt.Errorf("could not read values from source: %w", err)
+	}
+
+	return chartutil.Values(chartutil.CoalesceTables(d, s)).YAML()
+}
+
+func (c *component) generateLinkerDCerts() error {
+	root, err := tls.GenerateRootCAWithDefaults("identity.linkerd.cluster.local")
+	if err != nil {
+		return fmt.Errorf("generating certificates: %w", err)
+	}
+
+	c.Key = indent(root.Cred.EncodePrivateKeyPEM(), 8)
```

Comment:
It would be better to return a struct here, rather than mutating the component struct.

Reply:
Do you mean like generate the struct here and return?

Reply:
Yes, do something like:
```go
return &certs{
  key: indent(root.Cred.EncodePrivateKeyPEM(), 8)
  ...
}
```

Then assign it after calling this function.

Reply:
But then I will have to add a new struct for that. But for now there is no need to add one right? Are there any issues you see with mutating that won't be encountered when returning?

Reply:
As I mentioned earlier, I think having a new struct here is fine, as it nicely groups those generated values and makes this function simpler.

Reply:
added a new struct.